### PR TITLE
Implementation of Community object init and factory methods.

### LINKED ISSUE #278: Implementation of Community object
This issue is to create a PR to provide an implementation of the Community class. It is intended to isolate that particular component of @AmyOctoCat's work in progress PR https://github.com/ImperialCollegeLondon/pyrealm/pull/231 to allow a review and refactor of that section of the code and allow us to bring in that code in smaller chunks.

This issue should:

* provide something close to a final definition of the Community building on @AmyOctoCat's work and review by @davidorme,
* methods to create instances from data in files - and also handle multiple communities in files.
* test that code.
* provide API docs

Wider user docs can be delayed until more of the system is in place.

### PR DESCRIPTION
**Pull request by davidorme**
# Description

This PR brings in a Community object based of @AmyOctoCat's work in PR #230. From the features listed in #278, we have a (mostly) complete implementation with some elements postponed for later PRs:

- [x] provide something close to a final definition of the Community building on @AmyOctoCat's work and review by @davidorme,

> [!NOTE]
>   - The community object takes cohort data as arrays, merges in PFT traits for those arrays and then calculates the cored predictions of the T Model for the initialisation data.
>   - I've switched to using a `pandas.DataFrame` for storing cohort data.  I think that reverses something I told @AmyOctoCat (sorry!), but although there is overhead in using the Dataframe rather than simple arrays as attributes, it does simplify the namespace of the class a lot and it is also much cleaner to add and delete cohorts as rows rather than having to iterate over individual attributes.
  
- [ ] methods to create instances from data in files - and also handle multiple communities in files.

> [!NOTE]
>   - This **does not** implement loading multiple `Community` objects yet - might as well sense check the single case before doing that!
>   - The serialisation formats for CSV, JSON and TOML are revised
>   - Factory methods are now implemented to load from those formats.
>   - Marshmallow schemas are used to validate and post-process all of those formats into the format needed to initialise a `Community` object.

- [x] test that code.

> [!NOTE]
>  - It currently tests the T Model functions code rather lightly and it would be better to have deeper regression tests against the R implementation. I'd like to break that out into a new PR as complete tests would include some elements missing from the T model  functions at the moment and I don't want to bloat this basic Community PR.
  
- [x] provide API docs 

This PR also updates `Flora` to provide a `pandas.Dataframe` view of the flora data, which makes it super easy to merge onto the cohort data. 

There's some complexity in the serialisation formats and schemas:

- the `Community` object needs single values for `cell_area` and `cell_id` but arrays for the cohort data.
- CSV is great for rows of cohort data, but isn't structured for the single values, so I've imposed those as needing to be constant fields in the CSV. That is easy for users to maintain and feeds forward into having multiple cells in a file.
- JSON/TOML allows structured data, but editing arrays of cohort data is ugly and confusing because they're not neatly aligned in grids. So here, I've made the cohort data a list of objects, which is much easier to maintain in these formats.

The validation schemas take these two slightly different approaches, validate them and then coerce to the common arguments used for initialisation.

Fixes #278

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Key checklist

- [x] Make sure you've run the `pre-commit` checks: `$ pre-commit run -a`
- [x] All tests pass: `$ poetry run pytest`

## Further checks

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by davidorme** [comment]
There is still a doc failure - `sphinx` is failing to link to pandas DataFrame from the `flora.data` attribute. It is linking other refs to `pandas.DataFrame` but not here. No idea why.

We could just add this to the growing pile of noise in our `nitpick_ignore` section in the `sphinx` `conf.py` but I think there's something wrong in the setup - we surely shouldn't be getting this much trouble from `intersphinx` and `numpy`?

File: pyrealm/demography/community.py [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+(or 'cell') with a given cell id number and area.
+
+The class provides factory methods to create Community instances from CSV, JSON and TOML
+files, using :mod:`marshmallow` schemas to both validate the input data and to perform
+post processing to align the input formats to the initialisation arguments to the
+Community class.
+
+Internally, the cohort data in the Community class is represented as a pandas dataframe,
+which makes it possible to update cohort attributes in parallel across all cohorts but
+also provide a clean interface for adding and removing cohorts to a Community.
```
Comment (j-emberton) [inline pyrealm/demography/community.py:12]:
```suggestion
also provide a clean interface for adding and removing cohorts to a Community.

Worked example:
>>> from flora import PlantFunctionalType, Flora
>>> from t_model import (
...     calculate_heights, calculate_crown_areas, calculate_stem_masses,
...     calculate_foliage_masses
... )
>>> import pandas as pd

>>> pft1 = PlantFunctionalType(
    ...     name="Evergreen Tree",
    ...     a_hd=120.0,
    ...     ca_ratio=380.0,
    ...     h_max=30.0,
    ...     rho_s=210.0,
    ...     lai=3.0,
    ...     sla=12.0,
    ...     tau_f=5.0,
    ...     tau_r=1.2,
    ...     par_ext=0.6,
    ...     yld=0.65,
    ...     zeta=0.18,
    ...     resp_r=0.95,
    ...     resp_s=0.045,
    ...     resp_f=0.12,
    ...     m=2.5,
    ...     n=4.5,
    ... )

>>> pft2 = PlantFunctionalType(
    ...     name="Deciduous Shrub",
    ...     a_hd=100.0,
    ...     ca_ratio=350.0,
    ...     h_max=4.0,
    ...     rho_s=180.0,
    ...     lai=2.0,
    ...     sla=15.0,
    ...     tau_f=3.0,
    ...     tau_r=0.8,
    ...     par_ext=0.4,
    ...     yld=0.55,
    ...     zeta=0.15,
    ...     resp_r=0.85,
    ...     resp_s=0.05,
    ...     resp_f=0.1,
    ...     m=3.0,
    ...     n=5.0,
    ... )

    >>> pft3 = PlantFunctionalType(
    ...     name="Grass",
    ...     a_hd=80.0,
    ...     ca_ratio=300.0,
    ...     h_max=1.0,
    ...     rho_s=150.0,
    ...     lai=1.5,
    ...     sla=25.0,
    ...     tau_f=1.0,
    ...     tau_r=0.5,
    ...     par_ext=0.3,
    ...     yld=0.5,
    ...     zeta=0.12,
    ...     resp_r=0.8,
    ...     resp_s=0.04,
    ...     resp_f=0.09,
    ...     m=1.5,
    ...     n=3.5,
    ... )

>>> #Create a Flora collection and display the data:

>>> flora = Flora([pft1, pft2, pft3])

>>> #Example cohort data representing plant density, age, and associated PFTs:
>>> >>> cohorts = pd.DataFrame({
    ...     "cohort_id": [1, 2, 3, 4],
    ...     "pft_name": ["Evergreen Tree", "Deciduous Shrub", "Grass", "Evergreen Tree"],
    ...     "density": [100, 200, 500, 150],  # Number of individuals per unit area
    ...     "age": [5, 3, 2, 10],  # Age of the cohort in years
    ... })

>>> # Merge cohorts with flora data based on pft
>>> cohort_data = cohorts.merge(flora.data, left_on="pft_name", right_on="name")

>>> #Calculate tree height, crown area, stem mass, and foliage mass using T Model functions:

>>> cohort_data['height'] = calculate_heights(cohort_data['h_max'], cohort_data['a_hd'], cohort_data['dbh'])
>>> cohort_data['crown_area'] = calculate_crown_areas(cohort_data['ca_ratio'], cohort_data['a_hd'], cohort_data['dbh'], cohort_data['height'])
>>> cohort_data['stem_mass'] = calculate_stem_masses(cohort_data['rho_s'], cohort_data['height'], cohort_data['dbh'])
>>> cohort_data['foliage_mass'] = calculate_foliage_masses(cohort_data['sla'], cohort_data['lai'], cohort_data['crown_area'])

```

Comment (j-emberton) [reply pyrealm/demography/community.py:12]: I used chat gpt to help me get my head around the code. Feel free to incorporate, move or discard as you see fit.

File: pyrealm/demography/community.py [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+            isinstance(cohort_dbh_values, np.ndarray)
+            and isinstance(cohort_n_individuals, np.ndarray)
+            and isinstance(cohort_pft_names, np.ndarray)
+        ):
+            raise ValueError("Cohort data not passed as numpy arrays.")
+
+        # Check the cohort inputs are of equal length
+        if not (
+            (cohort_dbh_values.shape == cohort_n_individuals.shape)
+            and (cohort_dbh_values.shape == cohort_pft_names.shape)
```
Comment (j-emberton) [inline pyrealm/demography/community.py:300]: why not use pyrealm.core.utilities.check_input_shapes here?

File: pyrealm/demography/community.py [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+    #     :param csv_path: path to the csv containing community data, as detailed above.
+    #     :param flora: a flora object, ie a dictionary of plant functional properties,
+    #     keyed by pft name.
+    #     :return: a list of community objects, loaded from the csv
+    #     file.
+    #     """
+    #     community_data = pd.read_csv(csv_path)
+
+    #     data_grouped_by_community = community_data.groupby(community_data.cell_id)
+
```
Comment (j-emberton) [inline pyrealm/demography/community.py:499]: Is this surplus to requirements now?

Comment (davidorme) [reply pyrealm/demography/community.py:499]: No - but I need to think about a common interface to read multiple communities from CSV/TOML/JSON. Could be a class (`CommunityList`?), could be standalone functions. It's there so I don't forget to do that.

**Comment by j-emberton** [comment]
I've had a good poke around this code and it looks ok to my eye. I can't claim to be completely abreast of the science, but I like the fact you're rolling out this 'schema' approach that flows from what we discussed for the prior flora PR. The approach seems sensible.

I did I quick check and there is a GPU drop (cuDF) in for pandas. Hopefully this should mean that all this is compatible with the array api stuff I've been talking about.

Happy to approve once the

**Comment by j-emberton** [comment]
> There is still a doc failure - `sphinx` is failing to link to pandas DataFrame from the `flora.data` attribute. It is linking other refs to `pandas.DataFrame` but not here. No idea why.
> 
> We could just add this to the growing pile of noise in our `nitpick_ignore` section in the `sphinx` `conf.py` but I think there's something wrong in the setup - we surely shouldn't be getting this much trouble from `intersphinx` and `numpy`?

I haven't had time to dig into this issue at all I'm afraid. I'd hope we can fix it rather than ignore.

**Comment by davidorme** [comment]
@j-emberton Thanks for the worked example and other requests. Should all be good now. 

I have just `nitpick-ignored` the `intersphinx` issue with `pandas`. I'd much rather solve it properly, but this kind of thing always seems to turn into a huge time sink and I don't have the time now.

**Review by j-emberton — APPROVED** [review]
Following discussion and changes, happy to review

**Pull request merged**